app/models/report.rb

```diff
@@ -14,4 +18,33 @@ def editable?(target_user)
   def created_on
     created_at.to_date
   end
+
+  def fetch_mentioning_report_ids_diff
+    report_ids_in_content = extract_report_ids_in_content
+    mentioning_report_ids = mentioning_reports.map(&:id)
+
+    {
+      added_ids: report_ids_in_content - mentioning_report_ids,
+      removed_ids: mentioning_report_ids - report_ids_in_content
+    }
+  end
+
+  def add_mentions(added_ids)
+    added_ids.each { |id| active_mentions.create(mentioned_id: id) }
+  end
+
+  def remove_mentions(removed_ids)
+    active_mentions.where(mentioned_id: removed_ids).destroy_all
+  end
+
+  private
+
+  def extract_report_ids_in_content
+    urls_without_comment_url = search_urls_without_comment_url
+    urls_without_comment_url.join.scan(%r{reports/(\d+)}).flatten.map(&:to_i)
```

Comment:
```suggestion
    urls_without_comment_url.join.scan(%r{reports/(\d+)\b}).flatten.map(&:to_i)
```

みたいにすると48行目でcommentsを除外する処理が不要になりませんかね？

将来的に comments 以外のパターンが増える可能性もあるので、ブラックリスト方式でフィルタリングするのはバグを生みやすくなりそうです。

Reply:
> ```suggestion
>     urls_without_comment_url.join.scan(%r{reports/(\d+)\b}).flatten.map(&:to_i)
> ```
> 
> みたいにすると48行目でcommentsを除外する処理が不要になりませんかね？

正規表現`reports/(\d+)\b`に変更することで、48行目のcommentsを除外する処理が不要にならないと考えています！

Rubularで試してみた結果、
正規表現`reports/(\d+)\b`を使うと、コメントのURLの日報のidを取得するみたいで、
コメントのURLの日報のidを取得せずに、日報のURLのidを取得するようにならないみたいです。

コメントのURLに含まれる日報も言及することにすれば、48行目のcommentsを除外する処理は不要になるかと思います。
そのため、コメントのURLに含まれる日報も言及することにしようかと思います。

![スクリーンショット 2023-12-27 10 25 06](https://github.com/icchankun/fjord-books_app-2023/assets/82430737/98138106-d2bb-4f6c-8e3e-f3ac391f05c1)

> 将来的に comments 以外のパターンが増える可能性もあるので、ブラックリスト方式でフィルタリングするのはバグを生みやすくなりそうです。

そうですね。ブラックリスト方式でフィルタリングするのは、控えるようにします！